libraries/AP_HAL_PX4/CAN.h

```diff
@@ -0,0 +1,332 @@
+static const int16_t ErrBitRateNotDetected = 1007; ///< Auto bit rate detection could not be finished
+
+/**
+ * RX queue item.
+ * The application shall not use this directly.
+ */
+struct CanRxItem {
+    uint64_t utc_usec;
+    uavcan::CanFrame frame;
+    uavcan::CanIOFlags flags;
+    CanRxItem() :
+        utc_usec(0), flags(0)
+    {
+    }
+};
+
+struct CriticalSectionLocker {
+    const irqstate_t flags_;
+
+    CriticalSectionLocker() :
+        flags_(irqsave())
+    {
+    }
+
+    ~CriticalSectionLocker()
+    {
+        irqrestore(flags_);
+    }
+};
+
+namespace clock {
+uint64_t getUtcUSecFromCanInterrupt();
+uavcan::MonotonicTime getMonotonic();
+}
+
+class BusEvent: uavcan::Noncopyable {
+public:
+    BusEvent(PX4CANDriver& can_driver);
+    ~BusEvent();
+
+    bool wait(uavcan::MonotonicDuration duration);
+    static void signalFromCallOut(BusEvent *sem);
+
+    void signalFromInterrupt();
+    sem_t _wait_semaphore;
+    volatile uint16_t _signal;
+};
+
+class PX4CAN: public AP_HAL::CAN {
+    struct Timings {
+        uint16_t prescaler;
+        uint8_t sjw;
+        uint8_t bs1;
+        uint8_t bs2;
+
+        Timings() :
+            prescaler(0), sjw(0), bs1(0), bs2(0)
+        {
+        }
+    };
+
+    struct TxItem {
+        uavcan::MonotonicTime deadline;
+        uavcan::CanFrame frame;
+
+        bool pending;
+
+        bool loopback;
+
+        bool abort_on_error;
+
+        TxItem() :
+            pending(false), loopback(false), abort_on_error(false)
+        {
+        }
+    };
+
+    enum {
+        NumTxMailboxes = 3
+    };
+    enum {
+        NumFilters = 28
+    };
+
+    static const uint32_t TSR_ABRQx[NumTxMailboxes];
+
+    ObjectBuffer<CanRxItem> rx_queue_;
+    bxcan::CanType* const can_;
+    uint64_t error_cnt_;
+    uint32_t served_aborts_cnt_;
+    BusEvent& update_event_;
+    TxItem pending_tx_[NumTxMailboxes];
+    uint8_t peak_tx_mailbox_index_;
+    const uint8_t self_index_;
+
+    bool had_activity_;
+
+    uint32_t bitrate_;
+
+    int computeTimings(uint32_t target_bitrate, Timings& out_timings);
+
+    virtual int16_t send(const uavcan::CanFrame& frame, uavcan::MonotonicTime tx_deadline,
+                         uavcan::CanIOFlags flags);
+
+    virtual int16_t receive(uavcan::CanFrame& out_frame, uavcan::MonotonicTime& out_ts_monotonic,
+                            uavcan::UtcTime& out_ts_utc, uavcan::CanIOFlags& out_flags);
+
+    virtual int16_t configureFilters(const uavcan::CanFilterConfig* filter_configs, uint16_t num_configs);
+
+    virtual uint16_t getNumFilters() const
+    {
+        return NumFilters;
+    }
+
+    void handleTxMailboxInterrupt(uint8_t mailbox_index, bool txok, uint64_t utc_usec);
+
+    bool waitMsrINakBitStateChange(bool target_state);
+
+    bool initialized_;
+public:
+    enum {
+        MaxRxQueueCapacity = 254
+    };
+
+    enum OperatingMode {
+        NormalMode, SilentMode
```

Comment:
To clarify, `SilentMode` is needed for automatic CAN bit rate detection, and it is not used anywhere else.